### PR TITLE
[receiver/kafka] Add Azure Resource Logs support

### PR DESCRIPTION
**Pull request by cparkins**
**Description:** <Describe what has changed.>
Add a format to support the Azure Resource Logs format.

**Link to tracking Issue:** <Issue number if applicable>
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/18210

**Testing:** <Describe what testing was performed and which tests were added.>
Locally tested sending Azure Diagnostic Setting logs into Event Hub and connecting via the Kafka API.

**Documentation:** <Describe the documentation added.>
README updated to include new format for logs.

### CODE REVIEW

**Comment by atoulme** [comment]
I am ok with this, but would it make sense to reuse the code between azure event hub and kakfa receivers? We can do that in a subsequent PR if it's tricky.

**Comment by cparkins** [comment]
> I am ok with this, but would it make sense to reuse the code between azure event hub and kakfa receivers? We can do that in a subsequent PR if it's tricky.

I think that's a completely viable option.  Just not completely sure where to put the shared Unmarshaler to be honest. Once my changes for Event Hub are complete I could use those changes and they should be compatible with Kafka as well.

File: receiver/kafkareceiver/azureresourcelogs_unmarshaler.go [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+	var azureLogs azureRecords
+	decoder := jsoniter.NewDecoder(bytes.NewReader(buf))
+	if err := decoder.Decode(&azureLogs); err != nil {
+		return l, err
+	}
+
+	resourceLogs := l.ResourceLogs().AppendEmpty()
+	scopeLogs := resourceLogs.ScopeLogs().AppendEmpty()
+	scopeLogs.Scope().SetName(receiverScopeName)
+	scopeLogs.Scope().SetVersion("0.70.0")
```
Comment (MovieStoreGuy) [inline receiver/kafkareceiver/azureresourcelogs_unmarshaler.go:92]: Are you looking to set the collector build version? 

I believe you can fetch the version from `component.Host`, and you could just provided a constructor method:

```golang
func NewAzureEncoder(version string) (...)

// .... ignoring all the other magic needed

var encoders = map[string]encoders {
   `azureevents`: NewAzureEncoder(host.Version())
}
```

Not sure if it is that is, however, that would be how I would add in versions.

File: receiver/kafkareceiver/azureresourcelogs_unmarshaler.go [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+	logRecords := scopeLogs.LogRecords()
+
+	resourceID := ""
+	for _, azureLog := range azureLogs.Records {
+		resourceID = azureLog.ResourceID
+		nanos, err := asTimestamp(azureLog.Time)
+		if err != nil {
+			continue
```
Comment (MovieStoreGuy) [inline receiver/kafkareceiver/azureresourcelogs_unmarshaler.go:100]: Is it worth adding a debug statement here in the event a customer needs to debug this?

**Review by MovieStoreGuy — COMMENTED** [review]
This makes sense to me, there is a few things I would to get some of your thoughts on.

- How stable are the Azure logs? Do they frequently change their definitions?
- (I haven't really helped here) but how do you see backwards compatibility maintained for event definitions?

File: receiver/kafkareceiver/azureresourcelogs_unmarshaler.go [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	conventions "go.opentelemetry.io/collector/semconv/v1.13.0"
+)
+
+// Constants for OpenTelemetry Specs
+const receiverScopeName = "otelcol/" + typeStr
+
+// Constants for Azure Log Records
+const azureCategory = "azure.category"
```
[the remark this answers is not in the document]
Comment (cparkins) [reply receiver/kafkareceiver/azureresourcelogs_unmarshaler.go:33]: Not a problem.

File: receiver/kafkareceiver/README.md [hunk 4]
```diff
@@ -27,6 +27,7 @@ The following settings can be optionally configured:
   - `zipkin_proto`: the payload is deserialized into a list of Zipkin proto spans.
   - `zipkin_json`: the payload is deserialized into a list of Zipkin V2 JSON spans.
   - `zipkin_thrift`: the payload is deserialized into a list of Zipkin Thrift spans.
+  - `azureresourcelogs`: (logs only) the payload is deserialized into a list of Azure Resource Logs (JSON).
```
[the remark this answers is not in the document]
Comment (cparkins) [reply receiver/kafkareceiver/README.md:30]: I would expect it to behave similar to how the raw format behaves.

File: receiver/kafkareceiver/azureresourcelogs_unmarshaler.go (hunk 1 above)
Comment (cparkins) [reply receiver/kafkareceiver/azureresourcelogs_unmarshaler.go:92]: Yes, this was left by accident when testing. I will update.

File: receiver/kafkareceiver/azureresourcelogs_unmarshaler.go (hunk 2 above)
Comment (cparkins) [reply receiver/kafkareceiver/azureresourcelogs_unmarshaler.go:100]: Yes, I added it to the Event Hub Unmarshaler already in that PR.

**Comment by atoulme** [comment]
Please rebase

File: receiver/kafkareceiver/azureresourcelogs_unmarshaler.go [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+	OperationName     string       `json:"operationName"`
+	OperationVersion  *string      `json:"operationVersion"`
+	Category          string       `json:"category"`
+	ResultType        *string      `json:"resultType"`
+	ResultSignature   *string      `json:"resultSignature"`
+	ResultDescription *string      `json:"resultDescription"`
+	DurationMs        *string      `json:"durationMs"`
+	CallerIPAddress   *string      `json:"callerIpAddress"`
+	CorrelationID     *string      `json:"correlationId"`
+	Identity          *interface{} `json:"identity"`
```
Comment (MovieStoreGuy) [inline receiver/kafkareceiver/azureresourcelogs_unmarshaler.go:75]: Any reason to leave this as an `interface` ? would `[]byte` be better? (sorry, just trying to understand the choice made.

**Review by MovieStoreGuy — APPROVED** [review]

**Comment by MovieStoreGuy** [comment]
> @MovieStoreGuy I believe everything is ready for review. It appears that the build fail is not from my changes but I don't know how to re-queue other than submit changes (which I have but aren't ready).

Be sure to rebase against main as well, I have found that often helps in most cases.

**Pull request closed without merging**